optimum/habana/transformers/trainer.py

```diff
@@ -2593,91 +2581,44 @@ def _zero_model_grad(self, model):
 
-    def get_num_items_in_batches(
-        self,
-        args,
-        epochs_trained,
-        num_train_epochs,
-        train_dataloader,
-        len_dataloader,
-        num_examples,
-        steps_trained_in_current_epoch,
-    ):
-        """
-        Calculate the number of items in each batch for all epochs during training.
-        """
-        steps_in_epoch = (
-            len_dataloader if len_dataloader is not None else args.max_steps * args.gradient_accumulation_steps
+    def get_batch_samples(self, epoch_iterator, num_batches):
+        batch_samples = []
+        num_items_in_batch = None
+        count_num_items_in_batch = (
+            # num_items_in_batch is passed to model forward
+            # https://github.com/huggingface/transformers/blob/v4.49.0/src/transformers/trainer.py#L3757
+            self.model_accepts_loss_kwargs
+            # num_items_in_batch is passed to compute_loss_func
+            # https://github.com/huggingface/transformers/blob/v4.49.0/src/transformers/trainer.py#L3773
+            or self.compute_loss_func is not None
+            # num_items_in_batch is also verified if (self.model_accepts_loss_kwargs or self.compute_loss_func)
+            # https://github.com/huggingface/transformers/blob/v4.49.0/src/transformers/trainer.py#L3790
         )
 
-        remainder = num_examples % args.gradient_accumulation_steps
-        if remainder == 0:
-            remainder = args.gradient_accumulation_steps
-
-        total_updates = steps_in_epoch // args.gradient_accumulation_steps + 1
-        if args.gradient_accumulation_steps == 1:
-            total_updates -= 1
-        global_step = 0
-
-        num_items_in_batches = []
-        for epoch in range(epochs_trained, num_train_epochs):
-            if epoch == epochs_trained and steps_trained_in_current_epoch > 0:
-                epoch_dataloader = skip_first_batches(train_dataloader, steps_trained_in_current_epoch)
-            else:
-                epoch_dataloader = train_dataloader
-
-            if hasattr(epoch_dataloader, "set_epoch"):
-                epoch_dataloader.set_epoch(epoch)
-
-            epoch_iterator = iter(epoch_dataloader)
+        for _ in range(num_batches):
             try:
-                first_batch = next(epoch_iterator)
-            except StopIteration:
-                break
-            # Check if the batch contains "labels" (once per epoch)
-            if "labels" not in first_batch:
-                num_items_in_batches.append([None] * total_updates)
-                continue
-
-            device = first_batch["labels"].device
-
-            # Reset the iterator
-            epoch_iterator = iter(epoch_dataloader)
-
-            num_items_in_batches.append([])
-            for update_step in range(total_updates):
-                num_batches = args.gradient_accumulation_steps if update_step != (total_updates - 1) else remainder
+                batch = next(epoch_iterator)
```

Comment:
Let's have first 
```py
try:
    batch = next(epoch_iterator)
except StopIteration:
    break
```
and then the block
```py
if count_num_items_in_batch:
    try:
        ...
    except ...:
        ...
```
That way the code is a bit easier to compare to Transformers (even though that's a change you should probably do in Transformers IMO) and we avoid having two nested try-except block.

WDYT?

Reply:
Yes that works too, I actually changed many things compared to transformers on the journey to optimize this function xD
In this one there's also one loop vs two decoupled loops in transformers (one is list comprehension).
at some point I even pre-allocated `batch_samples = [None] * num_batches` 😭